single_kernel_mongo/managers/cluster.py

```diff
@@ -0,0 +1,371 @@
+                "mongos is integrated to a different CA than the config server. Please use the same CA for all cluster components."
+            )
+        if self.state.upgrade_in_progress:
+            raise DeferrableError(
+                "Processing client applications is not supported during an upgrade. The charm may be in a broken, unrecoverable state."
+            )
+
+    def relation_created(self) -> None:
+        """Just sets a status on relation created."""
+        logger.info("Integrating to config-server")
+        self.charm.status_manager.to_waiting("Connecting to config-server")
+
+    def on_database_created(self, username: str | None, password: str | None):
+        """Database created event.
+
+        Stores credentials in secrets and share it with clients.
+        """
+        if not username or not password:
+            raise WaitingForSecretsError
+        if self.state.upgrade_in_progress:
+            logger.warning(
+                "Processing client applications is not supported during an upgrade. The charm may be in a broken, unrecoverable state."
+            )
+            raise DeferrableFailedHookChecksError
+
+        if not self.charm.unit.is_leader():
+            return
+
+        logger.info("Database and user created for mongos application.")
+        self.state.secrets.set(AppPeerDataKeys.username.value, username, Scope.APP)
+        self.state.secrets.set(AppPeerDataKeys.password.value, password, Scope.APP)
+
+    def relation_changed(self) -> None:
+        """Start/restarts mongos with config server information."""
+        self.assert_pass_hook_checks()
+        key_file_contents = self.state.cluster.keyfile
+        config_server_db_uri = self.state.cluster.config_server_uri
+
+        if not key_file_contents or not config_server_db_uri:
+            raise WaitingForSecretsError("Waiting for keyfile or config server db uri")
+
+        updated_keyfile = self.dependent.update_keyfile(key_file_contents)
+        updated_config = self.dependent.update_config_server_db(config_server_db_uri)
+
+        if updated_keyfile or updated_config or not self.dependent.is_mongos_running():
+            logger.info("Restarting mongos with new secrets.")
+            self.charm.status_manager.to_maintenance("starting mongos")
+            self.dependent.restart_charm_services()
+
+            # Restart on highly loaded databases can be very slow (up to 10-20 minutes).
+            if not self.dependent.is_mongos_running():
+                logger.info("Mongos has not started yet, deferring")
+                self.charm.status_manager.to_waiting("Waiting for mongos to start")
+                raise DeferrableError
+
+            self.charm.status_manager.to_active("")
+
+        if self.charm.unit.is_leader():
+            self.state.app_peer_data.db_initialised = True
+
+        self.dependent.share_connection_info()
+
+    def relation_broken(self, relation: Relation):
```

Comment:
related to a previous comment, when reading a manager code separately from the charm / lifecycles, it's hard to understand from a business logic standpoint, I believe names such as "cleanup_....."

Reply:
I have renamed all those methods to more meaningful ones